core/src/main/kotlin/gateway/MasterGateway.kt

```diff
@@ -22,12 +29,44 @@ public interface MasterGateway {
 
     public val events: Flow<ShardEvent>
 
-    public suspend fun startWithConfig(configuration: GatewayConfiguration): Unit = coroutineScope {
-        gateways.entries.forEach { (shard, gateway) ->
-            val config = configuration.copy(shard = configuration.shard.copy(index = shard))
-            launch {
-                gateway.start(config)
+    @Deprecated(
+        "Specify maxConcurrency. It can be obtained by calling the Route.GatewayBotGet endpoint.",
+        ReplaceWith("this.startWithConfig(configuration, maxConcurrency = Int.MAX_VALUE)"),
+        level = WARNING,
+    )
+    public suspend fun startWithConfig(configuration: GatewayConfiguration): Unit =
+        startWithConfig(configuration, maxConcurrency = Int.MAX_VALUE)
+
+    public suspend fun startWithConfig(configuration: GatewayConfiguration, maxConcurrency: Int) {
+        require(maxConcurrency > 0) { "Invalid maxConcurrency: $maxConcurrency" }
+
+        // see https://discord.com/developers/docs/topics/gateway#sharding-max-concurrency
+        return coroutineScope {
+
+            var previousRateLimitKey = -1
+            val readyListeners = mutableListOf<Job>()
+
+            // sort gateways.entries to start shards in order
+            for ((shardId, gateway) in gateways.entries.sortedBy { it.key }) {
+                require(shardId >= 0) { "Negative shardId: $shardId" }
+
+                val rateLimitKey = shardId % maxConcurrency
+
+                if (rateLimitKey <= previousRateLimitKey) {
+                    readyListeners.joinAll() // wait until all gateways from last bucket are started
+                    readyListeners.clear()
+                }
+
+                // make sure we don't miss the event by executing until first suspension point before starting gateway
+                readyListeners += launch(start = UNDISPATCHED) { gateway.events.first { it is Ready } }
+
+                val config = configuration.copy(shard = configuration.shard.copy(index = shardId))
+                launch { gateway.start(config) }
+
+                previousRateLimitKey = rateLimitKey
```

Comment:
Should we use this instead of `previousRateLimitKey`:
```kotlin
var previousRateLimitBucket = -1
// ...
    val rateLimitBucket = shardId / maxConcurrency

    if (rateLimitBucket > previousRateLimitBucket) {
        // ...
    }
    // ...
    previousRateLimitBucket = rateLimitBucket
```
This would only behave differently in a situation like [this](https://discord.com/channels/613425648685547541/697489244649816084/1021565107949551676).

Reply:
For the test cases this would mean the `expectedBuckets` would change for ``fun `example 2 but without shards 15-30`()`` from `1` to `2` and for ``fun `randomly distributed shards`()`` from `3` to `4`

Reply:
no longer under consideration